articles/101_build_tool.md

```diff
@@ -233,6 +231,10 @@ The following uses cases should be satisfied by the unified build tool.
 The tool needs to be able to build ROS 1 workspaces which can already be built using `catkin_make_isolated` / `catkin_tools`.
 It is up to the implementation to decide if it only supports the standard CMake workflow or also the *custom devel space concept* of `catkin`.
 
+In ROS 2 the concept of the *devel space* has intentionally been removed.
+In the future it might be feasible to provide the concept of *symlinked installs* in ROS 1 to provide a similar benefit without the downsides.
+Therefore not supporting this concept in the universal build tool is considered a viable path forward.
```

Comment:
For ROS 1 or just ROS 2? If this includes ROS 1 (only thing that make sense from the text), then I don't feel that this was properly resolved.

Though I'm in favor of getting rid of the devel space in ROS 1, I think we cannot commit to doing that (and therefore not supporting it in the build tool) until we propose how to deprecate and then remove the devel space. In my opinion, It should still be a requirement that the default build tool for ROS 1 needs to support the devel space, at least until it is in the process of being removed.

Did others (other than @dirk-thomas) have a different conclusion based on our discussions?

Reply:
I completely agree with you points. Therefore I don't see `colcon` becoming the recommended tool for ROS 1 until that has been worked out.

Reply:
> For ROS 1 or just ROS 2?

`colcon` doesn't aim support the concept of a devel space - so that applies for ROS 1 as well as ROS 2.

Reply:
> Therefore I don't see colcon becoming the recommended tool for ROS 1 until that has been worked out.

Perhaps I'm the only one (still looking for others input here), but I feel that's an important thing to figure out now, because if the conclusion (at some later date) is that ROS 1 cannot live without the devel space and colon will never support it, I think that reality should captured here.

If we were to later surmise that "it's unacceptable for the default ROS 1 build tool to not support devel space", then we would have a situation which is at odds with the decision we have now.

I don't suppose we have to sort it out now, but it was a surprise to me when I realized it had not been resolved.

Reply:
Let's summarize the "facts" first then:

* The devel space requires ROS packages to do "special" logic (not native to CMake) to support the devel space correctly (e.g. different CMake extra files for devel and install).
* The fact that a package might only work in one of the cases (devel vs. install) if it the code is incorrect is surprising to users (since they might only use one of the two).
* The advantage of not copying resources can be achieved with symlinked installations which doesn't come with any of the above cons.
* The only disadvantage of using symlink install over devel space is that it requires packages to have install rules.

Based on these we decided in ROS 2 to not support the concept of devel space. Imo the same rational applies for a new build tool in ROS 1 (since the facts haven't changed). There would be an easy transition path: to add install rules if they are not already there.

That doesn't imply that we would ever remove the devel space logic from `catkin`. If users would like to continue using it they are fine to do it using the existing build tools. I don't think we will ever remove those since we don't want to break existing workflows. Instead we would just recommend something different. Users just shouldn't expect active development on the not anymore recommended build tools - but they should continue to work.

Reply:
I think you might be missing my point.

I'm not arguing for or against having the devel space or the build tool support for it.

Instead, my point is that the intended future state ought to be decided on, it sound like it has already been, and that in the implications section should say something like "therefore the default ROS 1 build tool will not support a devel space only workspaces, i.e. the default build tool will require install rules".

Reply:
I agree with @wjwwood. We need to have certainty about the intended final result.

I have no problems with the idea of ROS 1 eventually not supporting a devel space. I personally dislike the idea and much prefer ROS 2's approach to supporting a rapid development workspace.

However, the idea that ROS 1 could continue to work with both catkin and the new build tool sounds infeasible to me, if they have different requirements on the CMakeLists.txt files, which is what I think will be the case (please correct me if I'm wrong). Users of the new build tool would probably be fine, but as more and more packages support *only* the new build tool, those still using catkin for their work flow might find it increasingly difficult to stay that way. This seems like it would lead to friction.

Reply:
> my point is that the intended future state ought to be decided on, it sounds like it has already been

I didn't intend to decide anything for ROS 1 in this article. The idea was to describe how a path forward could look like. A decision about ROS 1 should be made in an REP and/or discourse discussion.

> the idea that ROS 1 could continue to work with both catkin and the new build tool sounds infeasible to me, if they have different requirements on the CMakeLists.txt files, which is what I think will be the case (please correct me if I'm wrong). Users of the new build tool would probably be fine, but as more and more packages support only the new build tool, those still using catkin for their work flow might find it increasingly difficult to stay that way. This seems like it would lead to friction.

I am not sure a new recommended build tool not supporting the devel space has a big impact here. As I mentioned in [this comment](https://github.com/ros2/design/pull/169#discussion_r182474264) this kind of problem already exists in ROS 1 today. Some packages do only work in the devel space and some do only work in the install space (since these packages haven't been developed with both in mind and the developers / maintainers / users only use one of the two). The choice of build tool doesn't change that. The only difference would be that recommending the new build tool without support for devel spaces will likely shift the ratio to more packages working in install space.

Reply:
Is there any data available on what the ratio is between one or the other style?

Reply:
> Is there any data available on what the ratio is between one or the other style?

Not that I know of.